### PR TITLE
transperent by design vqa model in opencog!

### PR DESCRIPTION
**Pull request by noskill**
Port of tbd vqa model to opencog
Source https://github.com/davidmascharka/tbd-nets

### CODE REVIEW

File: experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    :param inheritance_set: set
+        Inhertiance links constructed during conversion
+    :return: Tuple[ExecutionOutputLink, list, set]
+        program in form of execution output link, unprocessed part of the program, set of inheritance links
+    """
+    current, rest = commands[0], commands[1:]
+    if inheritance_set is None:
+        inheritance_set = set()
+    scene = atomspace.add_node(types.VariableNode, "$Scene")
+    if current.startswith('query'):
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py:192]: It is worth to add examples of the commands to make it clear how parsing works for people who are not familiar with TransparentByDesign

File: experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        var = atomspace.add_node(types.VariableNode, "$X")
+        concept = atomspace.add_node(types.ConceptNode, query_type)
+        inh_link = atomspace.add_link(types.InheritanceLink, [var, concept])
+        inheritance_set.add(inh_link)
+        # todo: use build link build_eval_link(atomspace, variable=var, category=concept, eval_link_sub=sub_prog)
+        link = build_filter(atomspace, concept, var, exec_out_sub=sub_prog)
+        return link, left, inheritance_set
+    elif current.startswith('scene'):
+        concept = atomspace.add_node(types.ConceptNode, "BoundingBox")
+        inh_link = atomspace.add_link(types.InheritanceLink, [scene, concept])
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py:206]: Is it correct that scene inherits bounding box?

File: experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    """
+    current, rest = commands[0], commands[1:]
+    if inheritance_set is None:
+        inheritance_set = set()
+    scene = atomspace.add_node(types.VariableNode, "$Scene")
+    if current.startswith('query'):
+        query_type = current.split('_')[-1]
+        sub_prog, left, inh = return_prog(atomspace, rest)
+        inheritance_set |= inh
+        var = atomspace.add_node(types.VariableNode, "$X")
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py:196]: On the first glance, all variables will have the same name ```$X``` will so all of them effectively will be same variable. Is it expected?

File: experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+def return_prog(atomspace, commands, inheritance_set=None):
+    """
+    Convert tbd program to atomese
+
+    :param atomspace: Atomspace
+    :param commands: list
+    :param inheritance_set: set
+        Inhertiance links constructed during conversion
+    :return: Tuple[ExecutionOutputLink, list, set]
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py:185]: It is not clear how second returned value "unprocessed part of the program" is used?

**Comment by vsbogd** [comment]
Some small example of TransparentByDesign program and result of its transformation to Atomspace representation would be really helpful to understand it.

File: experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py (hunk 1 above)
Comment (noskill) [reply experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py:192]: Done

File: experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py (hunk 2 above)
Comment (noskill) [reply experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py:206]: Yes, it does

File: experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py (hunk 3 above)
Comment (noskill) [reply experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py:196]: Yes, with current programs i expect the variable $X to be defined and used only in one place. I added an assert to verify that it is indeed true.

File: experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py (hunk 4 above)
Comment (noskill) [reply experiments/opencog/pattern_matcher_vqa/tbd_cog/tbd_helpers.py:185]: It is used in recursive calls to generate nested atoms

**Review by vsbogd — APPROVED** [review]

**Pull request merged**